Move android app initialization to custom application class

emmauss opened the pull request:
<!--- See CONTRIBUTING.md for general guidelines on contributions -->

## What does the pull request do?
Adds an Application class to the android backend and moving initialization to it. This ensures that no matter which activity is launched, the avalonia subsystem will be initialized with the correct App class.


## What is the current behavior?
App initialization on android only occurs when the main activity is launched. Any other activity launched before or without the main activity will use the default avalonia Application, thus styles and resources will not get loaded. I'm guessing this behaviour was added just to get the android backend up and running, and hasn't been fixed since.


## What is the updated/expected behavior with this PR?
Avalonia App is initialized once in a process, and any activity launched hosting an avalonia control will have access to app styles.


## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->


## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
All android projects targeting nightly or v12 will have to be updated. `AvaloniaMainActivity<TApp>` is removed in favor of adding an Application class that derives from the `AvaloniaAndroidApplication<TApp>` class.

## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->

emmauss committed:
move android app initialization to custom application class

emmauss committed:
fix tests

emmauss committed:
update api suppression

MrJul commented on src/Android/Avalonia.Android/AvaloniaAndroidApplication.cs:
```diff
@@ -0,0 +1,45 @@
+
+            var lifetime = new SingleViewLifetime();
+
+            (this as IAndroidApplication)!.Lifetime = lifetime;
```
Use `((IAndroidApplication)this)` instead of `as` + `!`.

emmauss replied:
Updated.

MrJul reviewed (commented):
Don't forget to add the breaking change to https://github.com/AvaloniaUI/Avalonia/wiki/v12-Breaking-Changes once the PR gets merged.

emmauss committed:
address comments

MrJul reviewed (approved):
LGTM!

kerams commented:
I've added `android:name=".Application"` to `application` in the manifest, and split `MainActivity` into

```fsharp
[<Application>]
type Application (javaRef, transfer) =
    inherit AvaloniaAndroidApplication<FabApplication> (javaRef, transfer)

    override _.CreateAppBuilder () =
        Main.configureLogging ()
        (Main.configure activated.Publish).UseAndroid ()

[<Activity(Label = "tst",
           Theme = "@style/MyTheme.Starting",
           LaunchMode = LaunchMode.SingleTop,
           MainLauncher = true,
           ConfigurationChanges = (ConfigChanges.Orientation ||| ConfigChanges.ScreenSize ||| ConfigChanges.UiMode))>]
type MainActivity () =
    inherit AvaloniaMainActivity ()

    override _.OnRequestPermissionsResult (c, p, r) =
        Platform.OnRequestPermissionsResult (c, p, r)
        base.OnRequestPermissionsResult (c, p, r)

    override this.OnCreate s =
        AndroidX.Core.SplashScreen.SplashScreen.InstallSplashScreen this |> ignore
        base.OnCreate s
        (this :> IAvaloniaActivity).Activated.Add activated.Trigger
        Platform.Init (this, s)
```

which as far as I can see mirrors changes to the samples in this PR. When I try to run this though, I get `Android.Runtime.JavaProxyThrowable: '[System.InvalidOperationException]: Invalid Arrange rectangle.'`. Am I missing something?

emmauss commented:
> I've added `android:name=".Application"` to `application` in the manifest, and split `MainActivity` into
> 
> ```fsharp
> [<Application>]
> type Application (javaRef, transfer) =
>     inherit AvaloniaAndroidApplication<FabApplication> (javaRef, transfer)
> 
>     override _.CreateAppBuilder () =
>         Main.configureLogging ()
>         (Main.configure activated.Publish).UseAndroid ()
> 
> [<Activity(Label = "tst",
>            Theme = "@style/MyTheme.Starting",
>            LaunchMode = LaunchMode.SingleTop,
>            MainLauncher = true,
>            ConfigurationChanges = (ConfigChanges.Orientation ||| ConfigChanges.ScreenSize ||| ConfigChanges.UiMode))>]
> type MainActivity () =
>     inherit AvaloniaMainActivity ()
> 
>     override _.OnRequestPermissionsResult (c, p, r) =
>         Platform.OnRequestPermissionsResult (c, p, r)
>         base.OnRequestPermissionsResult (c, p, r)
> 
>     override this.OnCreate s =
>         AndroidX.Core.SplashScreen.SplashScreen.InstallSplashScreen this |> ignore
>         base.OnCreate s
>         (this :> IAvaloniaActivity).Activated.Add activated.Trigger
>         Platform.Init (this, s)
> ```
> 
> which as far as I can see mirrors changes to the samples in this PR. When I try to run this though, I get `Android.Runtime.JavaProxyThrowable: '[System.InvalidOperationException]: Invalid Arrange rectangle.'`. Am I missing something?

Please post the full stack trace and a simple reproduction

kerams commented:
Never mind, I think https://github.com/AvaloniaUI/Avalonia/issues/18806 is the actual problem, because I tracked the error to a `TabControl` from Material.Styles, which uses a `UniformGrid` - if I remove one of the tabs, the exception disappears.

Nevertheless, it's bizarre that this started manifesting after moving to 12.0.999-cibuild0056440-alpha from the previous build 12.0.999-cibuild0056436-alpha.
